apps/api/src/app/workflows-v2/usecases/generate-preview/generate-preview.usecase.ts

```diff
@@ -111,10 +111,72 @@ export class GeneratePreviewUsecase {
     }
   }
 
-  /**
-   * Merges the payload schema into the variables schema to enable proper validation
-   * and sanitization of control values in the prepareAndValidateContentUsecase.
-   */
+  private async initializePreviewContext(command: GeneratePreviewCommand) {
+    const stepData = await this.getStepData(command);
+    const controlValues = command.generatePreviewRequestDto.controlValues || stepData.controls.values || {};
+    const workflow = await this.findWorkflow(command);
+    const payloadSchema = await this.buildPayloadSchema.execute(
+      BuildPayloadSchemaCommand.create({
+        environmentId: command.user.environmentId,
+        organizationId: command.user.organizationId,
+        userId: command.user._id,
+        workflowId: command.workflowIdOrInternalId,
+        controlValues,
+      })
+    );
+    const variableSchema = this.buildVariablesSchema(stepData.variables, payloadSchema);
+
+    return { stepData, controlValues, variableSchema, workflow };
+  }
+
+  private buildVariableExample(
+    controlValueForPreview: Record<string, unknown>,
+    workflow: WorkflowInternalResponseDto,
+    commandVariablesExample: PreviewPayload | undefined
+  ) {
+    const templateVars = this.extractTemplateVariables([controlValueForPreview]);
+
+    if (templateVars.length === 0) {
+      // eslint-disable-next-line no-console
+      console.log('No template variables found');
+    }
+
+    const variablesExample = pathsToObject(templateVars, {
+      valuePrefix: '{{',
+      valueSuffix: '}}',
+    });
+
+    const variablesExampleForPreview = this.buildVariablesExample(workflow, variablesExample, commandVariablesExample);
+
+    return variablesExampleForPreview;
+  }
+
+  private processControlValueVariables(schemaCompliantControls: Record<string, unknown>, variableSchema: any) {
+    const { validVariables, invalidVariables } = extractLiquidTemplateVariables(
+      JSON.stringify(schemaCompliantControls)
+    );
+
+    const { validVariables: validSchemaVariables, invalidVariables: invalidSchemaVariables } =
+      validateValidVariablesAgainstSchema(variableSchema, validVariables);
+
+    const resultVariables = {
+      valid: validSchemaVariables,
+      invalid: [...invalidVariables, ...invalidSchemaVariables],
+    };
+
+    return resultVariables;
```

Comment:
`resultVariables.invalid` will work well for generating issues based on the liquid errors.